### PR TITLE
✨(frontend) teacher courses search with query

### PR DESCRIPTION
**Pull request by rlecellier**
note: this branch is rebased on https://github.com/openfun/richie/pull/2308

## Purpose

Add search field on teacher courses listing pages

image

[Capture vidéo du 28-02-2024 15:47:50.webm](https://github.com/openfun/richie/assets/139848/99129b5e-5ccc-4429-8207-630c56a3c552)

### CODE REVIEW

**Review by jbpenrath — COMMENTED** [review]
The logic to retrieve user or organization course is broken.

Indeed, if I go on the dashboard, the user course are displayed then I click on university logo and here courses are not reloaded.


https://github.com/openfun/richie/assets/9265241/1d4fbe3a-e10c-4bab-b311-65607ff35826

**Comment by jbpenrath** [comment]
Feedbacks about the UX of this feature : 
- I think a CTA to clear filter is missing
- I wonder if we should not automatically trigger the search when user is typing (then throttle the search) Maybe this is not a good idea at first as there is no ES behind the scene

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### Added
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply CHANGELOG.md:11]: + is the search button really needed?

[the remark this answers is not in the document]
Comment (NathanVss) [reply CHANGELOG.md:11]: This gap feels odd when there is no "n results found", WDYT about reducing it?

Capture d’écran 2024-03-05 à 11 14 46

File: src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+const useTeacherCoursesSearch = () => {
+  const [searchParams, setSearchParams] = useSearchParams();
+  const [courseAndProductList, setCourseAndProductList] = useState<
+    (CourseListItem | CourseProductRelation)[]
+  >([]);
+  const [isInitialLoading, setIsInitialLoading] = useState(false);
+
+  const { data, isLoading, next, hasMore, count } = useCourseProductUnion({
+    query: searchParams.get('query') || undefined,
```
Comment (NathanVss) [inline src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx]: is `|| undefined` needed?

File: src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    perPage: 25,
+    productType: ProductType.CREDENTIAL,
+  });
+
+  useEffect(() => {
+    if (!data.length && isLoading) {
```
Comment (NathanVss) [inline src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx:30]: what if `data.length` is 0 ?

File: src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      setIsInitialLoading(true);
```
Comment (NathanVss) [inline src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx]: can't we set a `return` here?

File: src/frontend/js/widgets/Dashboard/components/SearchResultsCount/index.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+interface SearchResultsCountProps {
+  nbResults?: number;
+}
+
+const SearchResultsCount = ({ nbResults }: SearchResultsCountProps) => {
+  const [searchParams] = useSearchParams();
+  return (
+    <div
+      className={classNames('list__count-description', {
+        'list__count-description--no-results': !searchParams.get('query') || !nbResults,
```
Comment (NathanVss) [inline src/frontend/js/widgets/Dashboard/components/SearchResultsCount/index.tsx:23]: What is this class used for?

File: src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx [hunk 6]
```diff
@@ -1,43 +1,70 @@
...
-      setIsInitialLoading(true);
+      setIsNewSearchLoading(true);
     }
-    if (isInitialLoading && !isLoading) {
-      setIsInitialLoading(false);
-      setCourseAndProductList(data);
-    } else if (!isLoading && data.length > courseAndProductList?.length) {
-      setCourseAndProductList(data);
+
+    if (!isLoading) {
```
Comment (NathanVss) [inline src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx:34]: what about a guardian pattern to reduce nesting?


```
if (isLoading) {
    return;
}
```

File: src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx [hunk 7]
```diff
@@ -1,43 +1,70 @@
...
+    hasMore,
+    count: currentCount,
+  } = useCourseProductUnion({
+    query,
+    organizationId,
     perPage: 25,
     productType: ProductType.CREDENTIAL,
   });
 
   useEffect(() => {
```
Comment (NathanVss) [inline src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx:29]: This useEffect is edited in two commits

File: src/frontend/js/components/TeacherDashboardCourseList/index.tsx [hunk 8]
```diff
@@ -29,28 +29,34 @@ const messages = defineMessages({
 interface TeacherDashboardCourseListProps {
   titleTranslated?: string;
   organizationId?: string;
+  loadMore: () => void;
```
Comment (NathanVss) [inline src/frontend/js/components/TeacherDashboardCourseList/index.tsx:32]: I might be a bit 🍺 but I don't understand why you had to add all those `useCourseProductUnion` to props to bring them up?

**Review by NathanVss — APPROVED** [review]
I approve with few suggestions and feedbacks, but also mind this issue that I encountered on this branch: https://github.com/openfun/richie/issues/2317

File: CHANGELOG.md [hunk 9]
```diff
@@ -10,6 +10,7 @@ Versioning](https://semver.org/spec/v2.0.0.html).
...
 
+- Add search bar on teacher dashboard courses pages.
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply CHANGELOG.md:13]: It's either that or adding `Nullable` to `useCourseProductUnion` query props.
Both a equivalent.

File: src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx (hunk 3 above)
Comment (rlecellier) [reply src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx:30]: That the point, when `data.length` is 0 with `isLoading === true` then it mean that we've trigger a new research.

File: src/frontend/js/components/TeacherDashboardCourseList/index.tsx (hunk 8 above)
Comment (rlecellier) [reply src/frontend/js/components/TeacherDashboardCourseList/index.tsx:32]: How would you suggest that these information get given to `TeacherDashboardCourseList` from `TeacherDashboardOrganizationCourseLoader` and `TeacherDashboardCourseLoader` if i don't use props ?
calling `useCourseProductUnion` a second time here would not work.

File: src/frontend/js/widgets/Dashboard/components/SearchResultsCount/index.tsx (hunk 5 above)
Comment (rlecellier) [reply src/frontend/js/widgets/Dashboard/components/SearchResultsCount/index.tsx:23]: When no research is active, we don't display the counter.

File: src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx (hunk 2 above)
Comment (rlecellier) [reply src/frontend/js/hooks/useTeacherCoursesSearch/index.tsx]: It's either that or adding `Nullable` to `useCourseProductUnion` query props.
Both a equivalent.

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (rlecellier) [reply CHANGELOG.md:11]: i made some updates:
* hided the clear button when no research is active
* reduce searchBar margin bottom from 20px to 5px

File: src/frontend/js/components/TeacherDashboardCourseList/index.tsx (hunk 8 above)
Comment (NathanVss) [reply src/frontend/js/components/TeacherDashboardCourseList/index.tsx:32]: Why not using `useTeacherCoursesSearch` directly inside `TeacherDashboardCourseList`?

**Review by jbpenrath — APPROVED** [review]

File: src/frontend/js/components/TeacherDashboardCourseList/index.tsx (hunk 8 above)
Comment (rlecellier) [reply src/frontend/js/components/TeacherDashboardCourseList/index.tsx:32]: I need information from it.
`TeacherDashboardCourseList` handle the render of the list.
The layout handle everything around, title, searchbar, counter.

**Pull request merged**